File: repository/src/main/java/by/rekuts/travelagency/repository/impl/TourRepositoryImpl.java
package by.rekuts.travelagency.repository.impl;

import by.rekuts.travelagency.domain.Tour;
import by.rekuts.travelagency.repository.Specification;
import by.rekuts.travelagency.repository.TourRepository;
import by.rekuts.travelagency.repository.TourSpecification;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.criteria.Predicate;
import java.util.List;

@Repository
public class TourRepositoryImpl implements TourRepository {
    @PersistenceContext
    private EntityManager entityManager;

    @Override
    public void insert(Tour tour) {
        entityManager.persist(tour);
    }

    @Override
    public void update(Tour tour) {
        var dbTour = entityManager.find(Tour.class, tour.getId());
        dbTour.setPhoto(tour.getPhoto());
        dbTour.setDate(tour.getDate());
        dbTour.setDuration(tour.getDuration());
        dbTour.setDescription(tour.getDescription());
        dbTour.setCost(tour.getCost());
        dbTour.setTourType(tour.getTourType());
        dbTour.setHotel(tour.getHotel());
        dbTour.setCountry(tour.getCountry());
        dbTour.setUsers(tour.getUsers());
        entityManager.merge(dbTour);
    }

    @Override
    public void delete(int id) {
        entityManager.remove(entityManager.find(Tour.class, id));
    }

    @Override
    public List<Tour> getList(Specification specification) {
        final var tourSpecification = (TourSpecification) specification;
        var builder = entityManager.getCriteriaBuilder();
        var criteriaQuery = builder.createQuery(Tour.class);
        var root = criteriaQuery.from(Tour.class);
        List<Predicate> predicates;
        predicates = tourSpecification.getPredicates(root, builder);
        if (!predicates.isEmpty()) {
            criteriaQuery.where(
                    predicates.toArray(new Predicate[]{})
            );
        }
        return entityManager.createQuery(criteriaQuery).getResultList();
    }
}

File: repository/src/main/scala/by/rekuts/travelagency/repository/impl/TourRepositoryImpl.scala
package by.rekuts.travelagency.repository.impl

import java.util

import by.rekuts.travelagency.domain.Tour
import by.rekuts.travelagency.repository.{Specification, TourRepository, TourSpecification}
import javax.persistence.criteria.{CriteriaBuilder, CriteriaQuery, Predicate, Root}
import org.springframework.stereotype.Repository

@Repository
class TourRepositoryImpl extends TourRepository {

  override def update(tour: Tour): Unit = {
    val dbTour = entityManager.find(classOf[Tour], tour.id)
    dbTour.photo_$eq(tour.photo)
    dbTour.date_$eq(tour.date)
    dbTour.duration_$eq(tour.duration)
    dbTour.description_$eq(tour.description)
    dbTour.cost_$eq(tour.cost)
    dbTour.tourType_$eq(tour.tourType)
    dbTour.hotel_$eq(tour.hotel)
    dbTour.country_$eq(tour.country)
    dbTour.users_$eq(tour.users)
    entityManager.merge(dbTour)
  }

  override def delete(id: Int): Unit = entityManager.remove(entityManager.find(classOf[Tour], id))

  override def getList(specification: Specification[_]): util.List[Tour] = {
    val tourSpecification: TourSpecification = specification.asInstanceOf[TourSpecification]
    val builder: CriteriaBuilder = entityManager.getCriteriaBuilder
    val criteriaQuery: CriteriaQuery[Tour] = builder.createQuery(classOf[Tour])
    val root: Root[Tour] = criteriaQuery.from(classOf[Tour])
    val predicates = tourSpecification.getPredicates(root, builder)
    if (!predicates.isEmpty) {
      criteriaQuery.where(predicates.toArray[Predicate](new Array[Predicate](predicates.size())): _*)
    }
    entityManager.createQuery(criteriaQuery).getResultList
  }
}
